### ISSUE TITLE
12C: No invoiceable lines in sale order after sale with delivery canceled and revalidated

### ISSUE THREAD

**Issue by colmeirin** [odoo/odoo#44209]
Impacted versions:
12C

Steps to reproduce:
Clean database, install sale, inventory, Manufacturing, Invoicing.
Create Product; Product type: storable product, Invoicing policy: delivered quantities.
Create 2 more products same way.
Add BOM to first product with product 2 and 3 and type "kit"/phantom.
Create sales order with first product and validate it. It will create the delivery.
Cancel order, it will cancel order and delivery.
Set to draft order.
Confirm order, it will create the delivery again.
Validate delivery.
Press "Create invoice" in order and select "invoiceable lines" and click create invoices

Current behaviour:
The following error appears: "There is no invoiceable line. If a product has a Delivered quantities invoicing policy, please make sure that a quantity has been delivered."

This can be troublesome with a multiple product sale order because following the same steps but adding a "simple" product to the order will generate an invoice with only "simple" product ignoring the kit products also delivered and never invoiced.

Expected behaviour:
Invoice all delivered products not only "not kit" ones.

Video/Screenshot link (optional):

**Comment by pedrobaeza** [comment]
You are not delivering anything, so I don't get what is wrong...

**Comment by colmeirin** [comment]
The propper stock picking is created and validated with both products, kit ones and simple ones. but when creating the invoice, only simple ones are invoiced.

**Comment by pedrobaeza** [comment]
And if you don't cancel the first delivery everything works OK?

**Comment by colmeirin** [comment]
Yes. When cancelled and revalidated, only simple products work ok, bom products are ignored when creating the invoiced although both products are configured with the same invoice policy and propper quantities are validated at the picking

**Comment by pedrobaeza** [comment]
OK, not sure what is wrong for that happening, but I'm trusting for now and letting the experts in this part to talk.

**Comment by colmeirin** [comment]
I did some further investigation and I tested the same use case in odoo11 and odoo13 and in both versions the invoice is properly created with BOM products delivered.

Looking at the code seems like this logic has been changing between 11, 12 and 13.
This is a comparison of the code between 13 and 12 that I think is responsible for this behaviour

odoo13: [https://github.com/odoo/odoo/blob/97c09e1c77284836c7c46b4e70d4f12eb48ce91b/addons/sale_mrp/models/sale_mrp.py#L21](url)
odoo12: [https://github.com/odoo/odoo/blob/5e2bf3fbf9253b20dafbaff6a5508b53c095f438/addons/sale_mrp/models/sale_mrp.py#L12](url)

<img width="931" alt="Captura de pantalla 2020-01-30 a las 10 21 17" src="https://user-images.githubusercontent.com/5647643/73436798-743d3d00-434b-11ea-8201-1d2296f501a9.png">

At the 12 version it performs a check against ALL moves that only pases if all moves are done(a problem when cancelling a sale order because the picking is cancelled and recreated when validating so you will end with more than 1 picking and some cancelled and 1 done).

So seems like this might be a bug of odoo12 version or I'm missing something else.

Hope the experts can throw some light on the matter
